### ISSUE TITLE
Multiprocess Grading

### ISSUE THREAD

**Issue by LinuxMercedes** [redkyn/grader#21]
Sometimes my lab gets cold and I don't have a space heater under my desk. Multiprocess grading would help all my CPU cores bring about the heat death of the universe faster!

**Comment by michaelwisely** [comment]
I don't see why not. If I remember, grader uses a bunch of async junk. It shouldn't be hard to do.

The Python part wouldn't be parallel -- just concurrent. The student's assignments would be running in subprocesses anyway, so those should run in parallel.

**Comment by brhoades** [comment]
Output here would be a challenge... Logging output is currently done live--- would it make more sense for async to just output all at once and log when it starts/ends?

**Comment by brhoades** [comment]
Additionally--- isn't this _as simple as_ wrapping a up:

https://github.com/redkyn/grader/blob/master/grader/grader/commands/grade.py#L42-L43

With a coroutine lambda then waiting?

**Comment by michaelwisely** [comment]
@brhoades I believe it's roughly that simple. It just needs an event loop and some coroutines.

Hopefully, `asyncio` has a sane way to set a limit on the maximum number of concurrently running tasks. We don't need 100 (in the case of CS 1001) containers trying to run `wget` at the same time... Maybe a flag (`grader grade hw -j4`) flag would be appropriate for limiting the number of concurrently graded assignments.

**Comment by brhoades** [comment]
Yep, those are my thoughts too. Default concurrency amount can be read from a gradesheet but defaults to the number of cores or something. I'll take a look at this due to its apparent simplicity sometime soon.
